includes/class-buoy-alert.php

```diff
@@ -1038,24 +1062,36 @@ public static function enqueueFrontEndScripts () {
         wp_register_script(
             self::$prefix.'-script',
             plugins_url(self::$prefix.'.js', dirname(__FILE__)),
-            array('jquery'),
+            array(
+                'jquery',
+                'underscore',
+                'backbone',
+            ),
             $plugin_data['Version']
         );
-        wp_localize_script(self::$prefix.'-script', self::$prefix.'_vars', self::localizeScript());
+        // Localize it.
+        wp_localize_script(self::$prefix.'-script', self::$prefix.'_vars', self::localizeUiText());
+        wp_localize_script(self::$prefix.'-script', self::$prefix.'_dom_hooks', self::$dom_hooks);
```

Comment:
> Would you mind explaining this for my wordpress education?

Sure. I'm going to take the opportunity to write up a more detailed explanation than you might have expected or maybe even need, consider it project documentation for future contributors if that's so.

> I was able to understand from the WP codex why the ui text would go through `wp_localize_text`, but I don't totally understand why the dom hooks thing is in here.

WordPress has only one mechanism in its API for pushing getting PHP-generated values into the JavaScript runtime of a page, and that mechanism is called `wp_localize_script()`. It was originally intended for client-side string interpolation in the context of localization, hence the name. The user's chosen locale, stored in the WordPress database, is queried by WordPress core, which then localizes all internationalized strings with its own localization API (built on GNU Gettext under the hood). But if these strings are needed by JavaScript, rather than simply being output to the page's HTML source, then they need to be injected into the JavaScript context somehow.

So the pipeline is:

    database value -> PHP value -> string interpolation (localization, in this case) -> pushed as global JavaScript object

WordPress also has its own asset queue API, which is responsible for managing dependencies between client-side assets such as JavaScripts and style sheets. To be used, each asset must first be registered with WordPress (by something like a call to [`wp_register_script()`](https://developer.wordpress.org/reference/functions/wp_register_script/)). "Registration," to WordPress, simply means telling WordPress about the URL of the script, giving it a name (called a "script handle," or more generally an "asset handle," in WordPress jargon) and optionally supplying additional metadata such as the version of the asset, any dependencies it has on other assets, and so on. Dependencies are listed by referring to the handle of the previously-registered asset. Here is [a list of all client-side assets and their handles shipped by WordPress core](https://developer.wordpress.org/reference/functions/wp_register_script/#core-registered-scripts).

This mechanism is [how we're actually including Backbone in the first place](https://github.com/betterangels/buoy/pull/204/commits/38e8217e8921e194fe14aa910ae0e43596beb810#diff-e55c04f9d2a229a6281815f2b43e76aaL1078).

Anyway, once an asset like a script is registered, it may get "enqueued," which means WordPress will take responsibility for actually adding it inside of a `<script>` element in the HTML page at some point. This usually happens by a call to [`wp_enqueue_script()`](https://developer.wordpress.org/reference/functions/wp_enqueue_script/). Sometimes a script (or other asset) is registered and enqueued in one go, rather than in separate steps.

For "localization," however—and, importantly, this includes *any* operation wherein the client-side JavaScript values are determined dynamically from the server-side PHP code—we first have to call `wp_localize_script()` ourselves. This is *kind of* like registering a new script, but actually behaves more like telling WordPress about even more metadata that it should associate with a given already-registered script. In this case, though, the additional metadata isn't version strings and so on, it's the dynamically generated values.

By calling `wp_localize_script()` and associating it with a given handle, WordPress will not only output the original script's source to the browser, it will *also* output a `<script>` element whose contents is a JavaScript representation of the PHP array that the `wp_localize_script()` was passed. Once that's done, the JavaScript source itself can access these dynamically-generated values in the JavaScript context by looking for a global whose name matches the second argument to `wp_localize_script()`.

TL;DR: Even though we're not technically "localizing" the `buoy_dom_hooks` strings, they are stored in PHP. Since they're stored in PHP, the cleanest way to move them into a JavaScript context is to use WordPress's localization API. The fact that the name of that function doesn't *really* match the use of that function *in this case* is an unfortunate consequence of WordPress's imperfect naming scheme and historical evolution.